wnfs/examples/tiered_blockstores.rs

```diff
@@ -0,0 +1,116 @@
+//! and storing actual file content in a block store that is meant to be 'cold' as in, reads only
+//! work with high latency.
+
+use anyhow::Result;
+use async_trait::async_trait;
+use chrono::Utc;
+use libipld::{Cid, IpldCodec};
+use rand::thread_rng;
+use std::{borrow::Cow, rc::Rc};
+use wnfs::private::{PrivateDirectory, PrivateForest, PrivateNode};
+use wnfs_common::{BlockStore, MemoryBlockStore};
+use wnfs_namefilter::Namefilter;
+
+#[async_std::main]
+async fn main() {
+    // Create a block store that holds all 'hot' data:
+    let mut hot_store = MemoryBlockStore::default();
+
+    // Create a block store that holds all 'cold' data.
+    // In reality this would probably be something that's accessible
+    // with very high latency, but with a lot of bandwidth & storage.
+    let mut cold_store = MemoryBlockStore::default();
+
+    // Create a random number generator for randomized encryption.
+    let rng = &mut thread_rng();
+
+    // Create a new private forest.
+    // This represents your whole private file system, but hides any internal structure.
+    let forest = &mut Rc::new(PrivateForest::new());
+
+    // Create a new private directory
+    let mut directory = Rc::new(PrivateDirectory::new(
+        Namefilter::default(),
+        Utc::now(),
+        rng,
+    ));
+
+    let file_path = ["datasets".into(), "recordings".into(), "monday.mp4".into()];
+    let video = b"This isn't actually a video. But it could be!";
+
+    // When 'opening' a file, we use the hot store to fetch any directories
+    // on the path we may have already created before.
+    let file = directory
+        .open_file_mut(&file_path, true, Utc::now(), forest, &mut hot_store, rng)
+        .await
+        .unwrap();
+
+    // `set_content` actually writes the data blocks to the blockstore in chunks,
+    // so for this we provide the `cold_store`.
+    file.set_content(Utc::now(), &video[..], forest, &mut cold_store, rng)
+        .await
+        .unwrap();
+
+    // When storing the hierarchy data blocks, we use the `hot_store`:
+    let private_ref = directory.store(forest, &mut hot_store, rng).await.unwrap();
+
+    // Same thing for the forest. Doing this will give us a single root CID
+    // for all of the data, but parts separated into `hot_store` and `cold_store`:
+    let private_root_cid = hot_store.put_async_serializable(forest).await.unwrap();
+
+    // We can now read out our data back:
+    let forest: Rc<PrivateForest> = Rc::new(
+        hot_store
+            .get_deserializable(&private_root_cid)
+            .await
+            .unwrap(),
+    );
+
+    let directory = PrivateNode::load(&private_ref, &forest, &hot_store)
+        .await
+        .unwrap()
+        .as_dir()
+        .unwrap();
+
+    // Reading the file's data will fail when only provided the hot store:
+    // assert!(directory.read(&file_path, true, &forest, &hot_store).await.is_err()); // currently panics for mysterious reasons?
```

Comment:
@appcypher can you help me understand why, when I uncomment this line, this causes a panic & stacktrace?
What I intended was simply to assert that the result of `.await` (without any `.unwrap()`) is an `Err`. But for some reason that doesn't work :thinking:

Reply:
Well, I *know* that there's an `Err`, and I *know* that the error is basically that a block is requested from `hot_store` which doesn't exist. But I *expect* that, and I just want to make sure with the assert that this is the case & document that.

Reply:
Ok, @appcypher and I figured this one out on a call. I submitted an issue: https://github.com/wnfs-wg/rs-wnfs/issues/228

Reply:
The issue is now fixed. Thanks @appcypher :)